### PR TITLE
Cortex-M: Use DWT for timing calculations' framework

### PR DESCRIPTION
**Pull request by ioannisg**
Currently rebased onto #27717

### CODE REVIEW

File: arch/arm/core/aarch32/cortex_m/CMakeLists.txt [hunk 1]
```diff
@@ -14,6 +14,7 @@ zephyr_library_sources(
   )
 
 zephyr_library_sources_ifdef(CONFIG_DEBUG_COREDUMP coredump.c)
+zephyr_library_sources_ifdef(CONFIG_CORTEX_M_DWT timing.c)
```
Comment (nashif) [inline arch/arm/core/aarch32/cortex_m/CMakeLists.txt:17]: I do not think this is correct, should depend on CONFIG_TIMING_FUNCTIONS instead.

File: arch/arm/core/aarch32/cortex_m/Kconfig [hunk 2]
```diff
@@ -288,6 +288,13 @@ config SW_VECTOR_RELAY_CLIENT
 	  sure the vector table pointer in RAM is set properly by the image upon
 	  initialization.
 
+config CORTEX_M_DWT
+	bool "Enable and use the DWT"
+	depends on CPU_CORTEX_M_HAS_DWT
+	default y if TIMING_FUNCTIONS
```
Comment (nashif) [inline arch/arm/core/aarch32/cortex_m/Kconfig:294]: I see that now, maybe doing this the other way around, where we enable  CORTEX_M_DWT if TIMING_FUNCTIONS...

File: arch/arm/core/aarch32/cortex_m/CMakeLists.txt (hunk 1 above)
Comment (ioannisg) [reply arch/arm/core/aarch32/cortex_m/CMakeLists.txt:17]: I wanted to
-  have a DWT-specific Kconfig switch to enable compiling the timing.c
- have a way to prioritize DWT's timing.c against SoC's timing.c (see later)

File: arch/arm/core/aarch32/cortex_m/Kconfig (hunk 2 above)
Comment (ioannisg) [reply arch/arm/core/aarch32/cortex_m/Kconfig:294]: Yes, good suggestion, will try that.

Comment (ioannisg) [reply arch/arm/core/aarch32/cortex_m/Kconfig:294]: fixed

**Pull request closed without merging**